pkg/wasm/cache.go

```diff
@@ -0,0 +1,184 @@
+// Copyright Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package wasm
+
+import (
+	"crypto/sha256"
+	"fmt"
+	"io/ioutil"
+	"net/url"
+	"os"
+	"path/filepath"
+	"sync"
+	"time"
+
+	"istio.io/pkg/log"
+)
+
+var wasmLog = log.RegisterScope("wasm", "", 0)
+
+const (
+	// DefaultWasmModulePurgeInteval is the default interval for periodic stale Wasm module clean up.
+	DefaultWasmModulePurgeInteval = 10 * time.Minute
+
+	// DefaultWasmModuleExpiry is the default duration for least recently touched Wasm module to become stale.
+	DefaultWasmModuleExpiry = 24 * time.Hour
+)
+
+// Cache models a Wasm module cache.
+// TODO(bianpengyuan): add metrics to the wasm package.
+type Cache interface {
+	Get(url, checksum string, timeout *time.Duration) (string, error)
+}
+
+// LocalFileCache for downloaded Wasm modules. Currently it stores the Wasm module as local file.
+type LocalFileCache struct {
+	// Map from Wasm module checksum to cache entry.
+	modules map[cacheKey]cacheEntry
```

Comment:
Maybe it should be keyed on SHA only? It would be nice not to require SHA in the config, and use something like e-tags to refresh resources on-demand.

Reply:
We can make SHA optional. Essentially we keep the existing cache key struct, If SHA is provided, we verify it, include that as part of key. If not, only url is used in the key.

> It would be nice not to require SHA in the config, and use something like e-tags to refresh resources on-demand.

Does you mean if sha is not provided, whenever a config update is pushed, we try downloading and see if it changes?

Reply:
Yes.

Reply:
Done. I added on demand refresh if checksum is not provided. I still keep cache key as combination of url and sha, it is just to prevent config error that downloading url updated without sha update, which I think could happen a lot.